### ISSUE TITLE
max-height value of "stretch" is missing from the listed Values

### ISSUE THREAD

**Issue by graymatter00** [mdn/content#29887]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/CSS/max-height

### What specific section or headline is this issue about?

Values

### What information was incorrect, unhelpful, or incomplete?

A description for the `stretch` value is missing.  However, it is mentioned as a part of the description for the `fit-content()` value, and it is in the **Browser compatibility** table.

### What did you expect to see?

An description for the `stretch` value.

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/css/max-height`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/max-height
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/css/max-height/index.md
* Last commit: https://github.com/mdn/content/commit/925b2bd8beeae6ce8237863637bcd28ccbb8d47f
* Document last modified: 2023-10-27T19:45:47.000Z

</details>

**Comment by hamishwillee** [comment]
This is valid, but for whoever does this it needs to also be added to these cases: width, height, inline-size, block-size, min-width, min-height, min-inline-size, min-block-size, max-width, max-height, max-inline-size, max-block-size - according to https://drafts.csswg.org/css-sizing-4/#sizing-values

@graymatter00 Are you interested in adding this?

**Comment by graymatter00** [comment]
Hi @hamishwillee. 

Good point on the need for the same requirement in the other size properties. 👍

> @graymatter00 Are you interested in adding this?

I don't think my documentation writing skills are up to the challenge.  I'll leave it to the good people who I'm sure can do a better job than me.

**Comment by hamishwillee** [comment]
No worries. @OnkarRuikar Just did some changes in this area and I'm hoping will jump on it.

**Comment by OnkarRuikar** [comment]
> No worries. @OnkarRuikar Just did some changes in this area and I'm hoping will jump on it.

I didn't document `stretch` and  `contain` properties while adding `fit-content` because no browser has implemented them. Though the BCD table shows `stretch` property implemented by chromium based browsers it hasn't actually been implemented by any browsers yet:
![1](https://github.com/mdn/content/assets/87750369/0d886be5-7ccd-45f2-96c4-28ff3f66624e)

The equivalent available value in chromium is `-webkit-fill-available` which doesn't seem to be well recognized.

I don't know if it is a good idea to document unimplemented experimental features because they tend to change a lot and not implemented. For example in case of `@page` at-rule [we had to remove a lot of documented stuff ](https://github.com/mdn/content/pull/28697 )which is there in specs but not implemented for a decade.

**Comment by hamishwillee** [comment]
@OnkarRuikar I knew you'd have a reason, but I saw it in the BCD. As you say, the rule is that we document stuff once there is an implementation. 

Either way, there is a bug here. I _suspect_ that perhaps the BCD reflects the version in which `stretch` can be used as an option to [fit-content](https://developer.mozilla.org/en-US/docs/Web/CSS/max-height#fit-content). If that's the case we need to rename that subfeature so it is more obvious.
Is there an easy way to test the assumption?

**Comment by OnkarRuikar** [comment]
> Either way, there is a bug here.

Yes, as per the BCD policy [Removal of irrelevant features](https://github.com/mdn/browser-compat-data/blob/main/docs/data-guidelines/index.md#removal-of-irrelevant-features) the `stretch` value shouldn't exist in BCD as no browser supports it. So the PR [Rename fill-available as stretch](https://github.com/mdn/browser-compat-data/pull/2842) shouldn't have landed and `-webkit-fill-available` should've been kept as it is. 

/cc @queengooborg

**Comment by hamishwillee** [comment]
Cool, but it's not just that. The docs reference it
![image](https://github.com/mdn/content/assets/5368500/23b10b21-d6ef-401c-9ce0-fcec172a1c19)

**Comment by dipikabh** [comment]
Hi @OnkarRuikar, I started looking at your PR https://github.com/mdn/content/pull/35262, and I thought I could continue the discussion here instead of in the PR.

I believe you had the correct idea in https://github.com/mdn/content/issues/29887#issuecomment-1785136602 to not document the value `stretch` yet; not only does it not have cross-browser support, but it is also not called `stretch` in Chrome or Firefox.  
- If we document `stretch` and inline-tag it as experimental (which you have done in your PR), I think it would be incomplete if we don't also mention in content not to use the name `stretch` since the supported version as per BCD has the name `-webkit-fill-available` (in case folks don't check the BCD table). So perhaps we should also inline tag it as non-standard, that is, if we document it? Probably a case to mark it as [`"standard_track": false`](https://github.com/mdn/browser-compat-data/blob/6ed580f9d6732b8f9c1a03093e2fdb4300fe5fca/css/properties/max-height.json#L346) in BCD as well.
BTW, I noticed Firefox's alternate name for it (`-moz-available`):

| Firefox | Chrome |
| -- | -- |
| <kbd><img src="https://github.com/user-attachments/assets/232f92b0-0cfd-4c83-b13b-4a3bc6a9762e" alt="max-height-firefox-128" /></kbd> | <kbd><img src="https://github.com/user-attachments/assets/6e117ced-6c7b-435d-8ee1-aa26541860f2" alt="max-height-chrome-127" /></kbd> |

- [BCD table for `width`](https://developer.mozilla.org/en-US/docs/Web/CSS/width#browser_compatibility) gives a clue about the alternate name for `stretch` in Firefox: `-moz-available`.
- I realize this particular issue has been for opened for the missing `stretch` value, but there's a similar problem with `anchor-size()`, in that it exists in the "Syntax" block and BCD table, but missing from "Values" (and "Formal syntax"). Just to draw a comparison between `anchor-size()` and `fit-content()`, since they are both experimental, only `fit-content()` appears in "Formal syntax". (btw, `anchor-size()` didn't work for me even in Chrome)

Looking at `max-height` via the lens of https://drafts.csswg.org/css-sizing/#max-size-properties (Level 3) and comparing with our content:
| value or keyword | "Syntax" section | "Values" section | "Formal syntax" section | Browser compat table |
| -- | -- | -- | -- | -- |
| `none`, `min-content`, `max-content` | ✅ | ✅ | ✅ | ✅ |
| `<length>` and `<percentage>` are fine although I would combine them as `<length-percentage>` under "Values" | ✅ | ✅ | ✅ | - |
| `fit-content(<length-percentage>)`  | ✅ should not be listed with keyword values | ✅ needs to be inline tagged experimental | ✅ | ✅ (experimental) |
| <calc-size()> |  | ❌ | ✅ | - |
| `anchor-size()` - not in the spec | ✅ example needs to be fixed to replace "height" | ❌  | ❌  | ✅ (experimental) |
| `fit-content` (Level 4)  | ✅ | ✅ | ❌ | ✅ |

Some other observations:
- <calc-size()> has not been covered anywhere in our docs
- Randomly looking at `block-size` to which we'll now be adding `stretch`, its [browser compat](https://developer.mozilla.org/en-US/docs/Web/CSS/block-size#browser_compatibility) table doesn't yet include `stretch`.
- [`stretch` in `height`'s BCD table](https://developer.mozilla.org/en-US/docs/Web/CSS/height#browser_compatibility) is not marked experimental, so there is some inconsistency there.
- Both the links in `max-height` "[Specifications](https://developer.mozilla.org/en-US/docs/Web/CSS/max-height#specifications)" resolve to the same [section in spec](https://drafts.csswg.org/css-sizing-4/#sizing-values), but perhaps it should be updated to point to https://drafts.csswg.org/css-sizing/#max-size-properties.

With the current discrepancies in browser compat data and "Formal syntax", my proposal is:
-  Not to document `stretch` until it reaches standardization. We can avoid using `stretch` in `fit-content`:
> `fit-content`: Uses the available space, but not more than `max-content`, i.e., min(max-content, max(min-content, `<length-percentage>`)), where `<length-percentage>` represents the available space.

and

> `fit-content([<length-percentage>])`: Uses the `fit-content` formula with the available space replaced by the specified argument, i.e. min(max-content, max(min-content, `<length-percentage>`)).
- BCD for `stretch` can be updated to non standard track.
- I don't know what's the story behind `anchor-size()`. I understand if we want to show it as an example where it can be passed into `calc()`, but I don't get it being listed as a separate entry in BCD table.

Let me know what you think.

Also /cc @hamishwillee again to get his inputs.

**Comment by OnkarRuikar** [comment]
> Not to document `stretch` until it reaches standardization. We can avoid using stretch in `fit-content`:

:+1:  Waiting for Hamish's opinion.

> BCD for stretch can be updated to non standard track.

The `stretch` value has been [documented in the specs](https://drafts.csswg.org/css-sizing-4/#valdef-width-stretch). So, how can we say it's non-standard?\
We could mention `-moz-available` and `-webkit-fill-available` in the value sections though iff we document the `stretch`.

> but missing from "Values" (and "Formal syntax"). 

The values not appearing in Formal syntax is an yari issue https://github.com/mdn/yari/issues/11607 .

> (btw, `anchor-size()` didn't work for me even in Chrome)  and but I don't get it being listed as a separate entry in BCD table

The `anchor-size()` value does work stand alone(without `calc()`) in Chromium 127 and 129:

https://github.com/user-attachments/assets/99ec3a25-c62c-46da-82a0-eb5a7f3fa36c

**Comment by dipikabh** [comment]
> The stretch value has been [documented in the specs](https://drafts.csswg.org/css-sizing-4/#valdef-width-stretch). So, how can we say it's non-standard?

True, by definition `standard_track` means the feature is part of an active specification, so in those terms BCD cannot be updated and I stand corrected. For [width](https://developer.mozilla.org/en-US/docs/Web/CSS/width#browser_compatibility), `stretch` is marked experimental even though it is supported in more than one browser engine. Perhaps in this case, "experimental" is also indicating that the feature has a non-standard name.

(While we are trying to move away from vendor-specific prefixes ([FAQ](https://developer.mozilla.org/en-US/docs/Learn/CSS/Howto/CSS_FAQ#what_do_the_-moz-_-ms-_-webkit-_-o-_and_-khtml-_properties_do)), browsers did not implement`stretch` behind a flag or preference (as is usually done for experimental features these days), but instead used prefixes.)

> The anchor-size() value does work stand alone(without calc()) in Chromium 127 and 129

I was trying it as a value in `max-height`. But yes, `max-height: anchor-size(width)`, `max-height: anchor-size(height)`, and `max-height: calc(anchor-size(height) * 2)` sure are working in Chrome.
(I don't know why I earlier tried `max-height: anchor-size(2em)`, which was invalid.)

Thanks for opening the yari issue!

**Comment by hamishwillee** [comment]
This is a complicated set of questions and also it looks like some of them have already been agreed by you guys. But on quick scan my opinion:

- `stretch` is standard, and appears to be applicable in some form or other in max-height, min-height, max-width, min-width, width
- We reflect BCD.
  - Syntax and Formal syntax for these cases need to mention `stretch` and remove alias.
  - `stretch` needs to be documented in those cases. Given the number of aliases I think it reasonable to have a note in those documents that this is commonly implemented as an alias, and they should check BCD. Very happy for the common aliases to be listed.
  - you might argue that it is also listing the aliases as properties, but I'd try avoid it.
- Other cases that don't use stretch or any alias of course omit it.
- Same logic should apply to `fit-content` etc.

In summary, "Always follow BCD but you can provide hints to non standard things in its docs".

Is that helpful? If not, let me know the issues which there is disagreement on?

**Comment by dipikabh** [comment]
Thanks, @hamishwillee.
"Always follow BCD but you can provide hints to non standard things in its docs" seems to be a good general guidance to run with it for this case.

Thankfully, Syntax and Formal syntax do not mention the aliases; they appear only in BCD.
(I found a mention of `-moz-available` on the [Mozilla Extensions](https://developer.mozilla.org/en-US/docs/Web/CSS/Mozilla_Extensions#width_min-width_and_max-width) page.)

@OnkarRuikar, when you have a moment, WDYT about making these updates to your PR:
- Removing `stretch` from properties that do not have it in their BCD (e.g., inline-size and block-size)
- Adding a note to the `stretch` description you've added to let users know to check browser compat for the alias used. I'm not too keen on explicitly naming them in the note, but rather just point users to check the BCD table. Let me know if you have other suggestions. 

If we're on the same page with the changes, I can proceed with reviewing the PR. Thanks!
